### PR TITLE
Use IP range set on task for Device NetInventory

### PR DESCRIPTION
**Pull request by eduardomozart**
Fix https://github.com/glpi-project/glpi-inventory-plugin/issues/523

### CODE REVIEW

**Review by stonebuzz — APPROVED** [review]
Hi @eduardomozart 

sounds OK to me (good job)

I don't have the means to test it, but the change doesn't seem risky.

Now that the problem has been clearly identified, we can revert if necessary

File: inc/networkinventory.class.php [hunk 1]
```diff
@@ -367,11 +367,33 @@ public function run($jobstate)
         $current = $jobstate;
         $agent->getFromDB($current->fields['agents_id']);
 
-        $ip = current(PluginGlpiinventoryToolbox::getIPforDevice(
+        $ips = current(PluginGlpiinventoryToolbox::getIPforDevice(
             $jobstate->fields['itemtype'],
             $jobstate->fields['items_id']
         ));
```
Comment (cedric-anne) [inline inc/networkinventory.class.php:373]: The loop `foreach ($ips as &$ip) {` will not work if you get the current element here.
```suggestion
        $device_ips = PluginGlpiinventoryToolbox::getIPforDevice(
            $jobstate->fields['itemtype'],
            $jobstate->fields['items_id']
        );
        $ip = current($ips);
```

File: inc/networkinventory.class.php [hunk 2]
```diff
@@ -367,11 +367,33 @@ public function run($jobstate)
...
+        foreach ($a_taskjobs as $a_taskjob) {
+            $a_definition = importArrayFromDB($a_taskjob['targets']);
+            foreach ($a_definition as $datas) {
+                $itemtype = key($datas);
+                $items_id = current($datas);
+    
+                switch ($itemtype) {
+                    case 'PluginGlpiinventoryIPRange':
+                        $pfIPRange->getFromDB($items_id);
+                        foreach ($ips as &$ip) {
```
Comment (cedric-anne) [inline inc/networkinventory.class.php:384]:
```suggestion
                        foreach ($device_ips as $device_ip) {
```

File: inc/networkinventory.class.php [hunk 3]
```diff
@@ -367,11 +367,33 @@ public function run($jobstate)
...
+                            if ($pfIPRange->getIp2long($ip) <= $pfIPRange->getIp2long($pfIPRange->fields['ip_end']) && $pfIPRange->getIp2long($pfIPRange->fields['ip_start']) <= $pfIPRange->getIp2long($ip)) {
+                                // in range, assign this IP as device IP
+                                break;
+                            }
+                        }
```
Comment (cedric-anne) [inline inc/networkinventory.class.php:389]:
```suggestion
                        foreach ($device_ips as $device_ip) {
                            if ($pfIPRange->getIp2long($device_ip) <= $pfIPRange->getIp2long($pfIPRange->fields['ip_end']) && $pfIPRange->getIp2long($pfIPRange->fields['ip_start']) <= $pfIPRange->getIp2long($device_ip)) {
                                // in range, assign this device IP
                                $id = $device_ip;
                                break;
                            }
                        }
```

File: inc/networkinventory.class.php [hunk 4]
```diff
@@ -367,11 +367,33 @@ public function run($jobstate)
...
+                        break;
+                        
+                }
+            }
+        }
+        if(!isset($ip)) $ip = current($ips);
```
Comment (cedric-anne) [inline inc/networkinventory.class.php:395]:
```suggestion
```

**Review by cedric-anne — CHANGES_REQUESTED** [review]

File: inc/networkinventory.class.php (hunk 4 above)
Comment (stonebuzz) [reply inc/networkinventory.class.php:395]: Why remove this fallback ? (if it's not a PluginGlpiinventoryIPRange or if the IPs don't match the Range)

Comment (cedric-anne) [reply inc/networkinventory.class.php:395]: Value is assigned outside the lopp in the first suggestion.
image

I think is is easier to understand and it prevents the usage of a reference in the loop.

**Review by cedric-anne — APPROVED** [review]

**Comment by eduardomozart** [comment]
Hello guys,
I did some changes on the script to handle cenarios where there's some sort of VIP (virtual IP, used for redundancy on network devices).
On my infrastructure, we have an Aruba Instant Access Points (IAP) cluster (a virtual controller that handle all APs on our site), as you can see below, both devices contains "192.168.0.6", which is a VIP that I use to manage the IAP cluster (only one device assume this IP at the time, and if this AP is down, other devices assume this VIP, so the cluster management is always available), but each AP has it's own AP:
image
image
So this virtual IP (192.168.0.6) is on the same range than my task (192.168.0.0/24), but it will be used only if there's no other unique IP available on the same network range.

**Comment by trasher** [comment]
This would need tests

**Comment by trasher** [comment]
Please fix CS

**Comment by eduardomozart** [comment]
> This would need tests

I'll see what I can do.

> Please fix CS

Which line(s)?

**Comment by trasher** [comment]
> > Please fix CS
> Which line(s)?

See GH actions output: https://github.com/glpi-project/glpi-inventory-plugin/actions/runs/9891523512/job/28031989775?pr=524

**Comment by eduardomozart** [comment]
Hello,
I added the Unit Tests as requested by trasher but as cedric already approved those changes and the changes I made after his approval was minimal (only continued loop if IP wasn't unique instead of breaking it), I believe it should work as expected.
The changes I made after cedric approval may not be needed (at least, for Aruba IAP cluster Virtual IPs) if #536 be investigated further, but maybe other systems of redundancy can also assign an VIP for more than one device, so this kind of check maybe useful to keep.

**Review by stonebuzz — APPROVED** [review]

**Review by trasher — APPROVED** [review]

**Pull request merged**